File: Code/infer_logx.jl
#using PyCall
#@pyimport matplotlib.pyplot as plt

# Load the files
scalars = transpose(readdlm("scalars.txt")) # Transpose for faster access
logw = vec(readdlm("logw.txt"))

# Truncate if necessary
N = minimum([size(scalars)[2], size(logw)[1]])
scalars = scalars[:, 1:N]
logw = logw[1:N]

# Which run each distribution belongs to
run_id = ones(Int64, N)
for i in 2:N
  if logw[i] == -1.
    run_id[i] = run_id[i-1] + 1
  else
    run_id[i] = run_id[i-1]
  end
end

# Generate logX values from the prior (i.e. not taking into account any
# inter-run comparisons)
logX = zeros(N)
for i in 1:N
  if logw[i] == -1.
    logX[i] = log(rand())
  else
    logX[i] = logX[i-1] + log(rand())
  end
end

# Gibbs sample one value
function update!(logx::Array{Float64, 1}, ii::Int64)
  # Generate proposal
  proposal = 0.

  # Range of values for proposal
  lower = -1E300
  upper = 0.

  # Lower limit -- next point in same run (if it exists)
  if (ii != N) && (run_id[ii+1] == run_id[ii])
    lower = logx[ii+1]
  end
  # Upper limit -- previous point in same run (if it exists)
  if (ii != 1) && (run_id[ii-1] == run_id[ii])
    upper = logx[ii-1]
  end

  # If lower limit exists, generate uniformly between limits
  if lower != -1E300
    proposal = lower + (upper - lower)*rand()
  else
    # Otherwise, use exponential distribution
    proposal = upper + log(rand())
  end

  # Measure inconsistency
  inconsistency_old = 0
  inconsistency_new = 0
  for j in 1:N
    if run_id[j] != run_id[ii]

      # Distribution ii is within distribution j
      if (logx[ii] > logx[j]) && all(scalars[:, ii] .>= scalars[:, j])
          inconsistency_old += 1
      end

      # Distribution ii is outside distribution j
      if (logx[ii] < logx[j]) && all(scalars[:, ii] .<= scalars[:, j])
	inconsistency_old += 1
      end

      # Distribution ii is within distribution j
      if (proposal > logx[j]) && all(scalars[:, ii] .>= scalars[:, j])
          inconsistency_new += 1
      end

      # Distribution ii is outside distribution j
      if (proposal < logx[j]) && all(scalars[:, ii] .<= scalars[:, j])
          inconsistency_new += 1
      end
    end
  end

  inconsistency = inconsistency_old
  # Accept?
  if inconsistency_new <= inconsistency_old
    logx[ii] = proposal
    inconsistency = inconsistency_new
  end
  return inconsistency
end

function update_all!(logx::Array{Float64, 1})
  total_inconsistency = 0
  for i in 1:N
    total_inconsistency += update!(logx, i)
  end
  return total_inconsistency
end

for i in 1:100
  inconsistency = update_all!(logX)
  println(i, " ", inconsistency)
end

writedlm("logX.txt", logX)

File: Code/infer_logx.cpp
#include <vector>
#include <fstream>
#include <iostream>
#include <cassert>
#include <cmath>
#include "RandomNumberGenerator.h"

using namespace std;
using namespace DNest3;

class Data
{
	private:
		vector<double> logw, run_id;
		vector< vector<double> > scalars;
		int N;

	public:
		Data()
		{ }

		void load()
		{
			logw.clear(); run_id.clear(); scalars.clear();

			double temp;
			fstream fin("logw.txt", ios::in);
			int k = 0;
			while(fin >> temp)
			{
				logw.push_back(temp);
				if(k == 0)
					run_id.push_back(0);
				else if(logw.back() == -1.)
					run_id.push_back(run_id.back() + 1);
				else
					run_id.push_back(run_id.back());
				k++;
			}
			fin.close();

			fin.open("scalars.txt", ios::in);
			double temp2;
			while(fin >> temp && fin >> temp2)
			{
				vector<double> vec(2);
				vec[0] = temp; vec[1] = temp2;
				scalars.push_back(vec);
			}
			fin.close();

			if(scalars.size() < logw.size())
			{
				logw.erase(logw.begin() + scalars.size(),
						logw.end());
				run_id.erase(run_id.begin() + scalars.size(),
						run_id.end());
			}
			if(logw.size() < scalars.size())
			{
				scalars.erase(scalars.begin() + logw.size(),
						scalars.end());
			}

			assert(logw.size() == scalars.size());
			assert(logw.size() == run_id.size());
			N = logw.size();
			cout<<"# Loaded "<<N<<" points."<<endl;
	}

	friend class Assignment;
};


/************************************************************************/

class Assignment
{
	private:
		const Data& data;
		vector<double> logX;

	public:
		Assignment(const Data& data)
		:data(data)
		{
		}

		void initialise()
		{
			logX.assign(data.N, 0.);

			for(int i=0; i<data.N; i++)
			{
				if(data.logw[i] == -1.)
					logX[i] = log(randomU());
				else
					logX[i] = logX[i-1] + log(randomU());			
			}
		}

		int update_all()
		{
			int total = 0;
			for(int i=0; i<data.N; i++)
			{
				total += update(i);
				cout<<"."<<flush;
			}
			cout<<endl;
			return total;
		}

		// Gibbs sample one value
		int update(int i)
		{
			double proposal;

			// Range of values for proposal
			double lower = -1E300;
			double upper = 0.;

			// Lower limit -- next point in same run (if it exists)
  			if( (i != (data.N-1)) &&
					(data.run_id[i+1] == data.run_id[i]))
    				lower = logX[i+1];
			// Upper limit -- previous point in same run (if it exists)
  			if( (i != 0) &&
					(data.run_id[i-1] == data.run_id[i]))
    				upper = logX[i-1];

			// If lower limit exists, generate uniformly between limits
			if(lower != -1E300)
				proposal = lower + (upper - lower)*randomU();
			else // otherwise use exponential distribution
				proposal = upper + log(randomU());

			// Measure inconsistency
			int inconsistency_old = 0;
			int inconsistency_new = 0;

			bool outside, inside;
			for(int j=0; j<data.N; j++)
			{
				if(data.run_id[j] != data.run_id[i])
				{
					// See if distribution j is outside,
					// inside, or unknown
					outside = true;
					inside  = true;
					for(int k=0; k<2; k++)
					{
						if(data.scalars[j][k] > data.scalars[i][k])
							outside = false;
						if(data.scalars[j][k] < data.scalars[i][k])
							inside = false;
					}

					if(outside && (logX[j] < logX[i]))
						inconsistency_old++;
					if(inside  && (logX[j] > logX[i]))
						inconsistency_old++;

					if(outside && (logX[j] < proposal))
						inconsistency_new++;
					if(inside  && (logX[j] > proposal))
						inconsistency_new++;
				}
			}

			int inconsistency = inconsistency_old;
			if(inconsistency_new <= inconsistency_old)
			{
				logX[i] = proposal;
				inconsistency = inconsistency_new;
			}
			return inconsistency;
		}

		void save()
		{
			fstream fout("logX.txt", ios::out);
			for(int i=0; i<data.N; i++)
				fout<<logX[i]<<endl;
			fout.close();
		}

};


int main()
{
	RandomNumberGenerator::initialise_instance();
	RandomNumberGenerator::get_instance().set_seed(time(0));

	Data data;
	data.load();

	Assignment assignment(data);
	assignment.initialise();

	int inconsistency;
	int when_finished = -1;
	int i=0;
	while(true)
	{
		inconsistency = assignment.update_all();
		assignment.save();
		if(when_finished == -1 && inconsistency == 0)
			when_finished = i;
		i++;
		cout<<i<<' '<<inconsistency<<endl;
		if(i > (1 + 1.2*when_finished))
			break;
	}
	return 0;
}

